python/ql/src/Functions/IterReturnsNonSelf.ql

```diff
@@ -4,27 +4,94 @@
  * @kind problem
  * @tags reliability
  *       correctness
+ *       quality
  * @problem.severity error
  * @sub-severity low
  * @precision high
  * @id py/iter-returns-non-self
  */
 
 import python
+import semmle.python.ApiGraphs
 
-Function iter_method(ClassValue t) { result = t.lookup("__iter__").(FunctionValue).getScope() }
+/** Holds if `f` is a method of the class `c`. */
+private predicate methodOfClass(Function f, Class c) {
+  exists(FunctionDef d | d.getDefinedFunction() = f and d.getScope() = c)
+}
+
+/** Gets the __iter__ method of `c`. */
+Function iterMethod(Class c) { methodOfClass(result, c) and result.getName() = "__iter__" }
+
+/** Gets the `__next__` method of `c`. */
+Function nextMethod(Class c) { methodOfClass(result, c) and result.getName() = "__next__" }
+
+/** Holds if `var` is a variable referring to the `self` parameter of `f`. */
+predicate isSelfVar(Function f, Name var) { var.getVariable() = f.getArg(0).(Name).getVariable() }
```

Comment:
I wonder, is the `.getVariable()` bit on both sides actually necessary, or would it suffice to do `var = f.getArg(0)`?

Reply:
I believe that would require that `var` is the parameter itself, rather than a use of it.

Reply:
Oh, you're right. This was a brain-fart on my part, sorry!